### PR TITLE
Make management of the printing environment safer

### PR DESCRIPTION
**Pull request by ghost**
As discussed in another PR, this PR tries to detect cases where `Printtyp.wrap_printing_env` should be used but is currently not.

I initially looked at @gasche's idea to add an abstract type for the printing environment and pass it around to the various functions. However, it seems to me that this is a big refactoring of not only `printtyp.ml` but also many other files in `typing/`. What I did instead is just add a runtime check to make sure we don't access the printing environment outside of `wrap_printing_env`. This was enough to find a few cases where this currently happens. In a couple of cases I had to add a dummy `wrap_printing_env Env.empty` to please the new check but that doesn't too bad.

The testsuite currently fails because some types are now printed with a `/2` suffix. I haven't found the reason yet, I'm just posting this PR in cases someone who knows this module better would have an idea of what's happening. For instance, this is one of the errors I get:

```
$ make one DIR=tests/typing-poly
Running tests from 'tests/typing-poly' ...
 ... testing 'poly.ml' with 1 (expect) => failed (program output poly.ml.corrected.corrected differs from reference poly.ml: 
--- poly.ml	2018-08-07 13:39:16.612843208 +0100
+++ poly.ml.corrected.corrected	2018-08-08 14:46:14.443472283 +0100
@@ -1152,9 +1152,9 @@
 module M : sig type 'a t val f : <m: 'a. 'a t> -> int end
 = struct type 'a t = int let f x = x#m end;;
 [%%expect {|
-module M : sig type 'a t type u = < m : 'a. 'a t > end
-module M : sig type 'a t val f : < m : 'a. 'a t > -> int end
-module M : sig type 'a t val f : < m : 'a. 'a t > -> int end
+module M : sig type 'a t type u = < m : 'a. 'a t/2 > end
+module M : sig type 'a t val f : < m : 'a. 'a t/2 > -> int end
+module M : sig type 'a t val f : < m : 'a. 'a t/2 > -> int end
 |}];;
 
 let f x y =
@@ -1410,8 +1410,8 @@
   sig
     type 'par t = 'par
     val ident : 'a -> 'a
-    class alias : object method alias : 'a t -> 'a end
-    val f : < m : 'a. 'a t > -> < m : 'a. 'a >
+    class alias : object method alias : 'a t/2 -> 'a end
+    val f : < m : 'a. 'a t/2 > -> < m : 'a. 'a >
   end
 |}];;
 

)
```

### CODE REVIEW

**Comment by Octachron** [comment]
This looks like the name collision mechanism picking another type `t` from the environment. I will have a look.

File: typing/printtyp.ml [hunk 1]
```diff
@@ -624,9 +673,8 @@ let same_printing_env env =
   Env.same_types !printing_old env && Concr.equal !printing_pers used_pers
 
 let set_printing_env env =
-  printing_env := env;
```
Comment (Octachron) [inline typing/printtyp.ml]: The root issue is here: both `-short-paths` and the name collision mechanism relied on the ability to update the environment when printing signatures. I think this function should be renamed to `recompute_map`, and the use of `set_printing_env` in `hide_rec_items`  and `still_in_type_group` should be replaced by call to `recompute_map+wrap_env`.

Fixing these two points (see the quick pach [here](https://github.com/Octachron/ocaml/commit/0918f1ded752aed9e1da3c8fafc872c8bc79f435)) let me with only the two failing tests in `tool-toplevel` and `tool-debugger` due to a missing_call to `wrap_printing_env`.

Comment (ghost) [reply typing/printtyp.ml]: Ah, I see. I didn't realize `set_printing_env` was called at all these places. The renaming you suggest makes sense to me. Thanks!

Comment (ghost) [reply typing/printtyp.ml]: I wasn't sure of the performance implications of adding more wrap_env, so I just made the behavior closer to the old code

**Comment by ghost** [comment]
This is now ready for review

**Comment by Octachron** [comment]
I am concerned about the `failwith` when the printing environment is unset: replacing an imperfect user-facing message error with an internal compiler error seems non-optimal outside of the compiler's test suite. Even more so, when the test suite coverage of error paths is imperfect.

**Comment by ghost** [comment]
That's true. However I claim that situation is improved by this PR. With master we get the following behavior when a `wrap_printing_env` is missing: the compiler reads more cmi files than it should. This results in weird build errors that are extremely hard to understand and debug.

With this PR, we immediately get an error. Users are not expected to understand it and it's likely that they'll simply report it upstream. An OCaml developer can then look at the reported backtrace and quickly fix the issue.

**Comment by Octachron** [comment]
Isn't that an argument for the typechecked path? Or maybe just discarding all lookup to `Env` if the printing env is not set. I think that at the very least, the error message need to be much more explict. And probably with a call `Misc.fatal_errorf` rather than a failwith.

There also some unprotected call to path in `toploop/topdirs.ml`:

```Ocaml
let f () = ();;
#trace f;;
#trace f;;
```
fails with
>Fatal error: exception Failure("Printtyp.Printing_env.get called outside of wrap_printing_env")

( [This commit](https://github.com/Octachron/ocaml/commit/53c2746373a0f01909b472bed9f78f7c4388f1c3) might be useful as a comparison with a overengineered but exhaustive first-class module version )

**Comment by ghost** [comment]
> ( This commit might be useful as a comparison with a overengineered but exhaustive first-class module version )

Your patch looks better indeed. Let's drop this PR then.

**Pull request closed without merging**